File: programmers30_43165/src/main/java/study/problem/Solution.java
package study.problem;

public class Solution {

	public int solution(int[] numbers, int target) {
		int answer = dfs(numbers, target, 0, 0);
		return answer;
	}

	public int dfs(int[] numbers, int target, int index, int sum) {
		if (numbers.length == index) {
			return sum == target ? 1 : 0;
		}
		return dfs(numbers, target, index + 1, sum + numbers[index])
			+ dfs(numbers, target, index + 1, sum - numbers[index]);
	}
}

File: programmers30_43165/src/main/kotlin/study/problem/kt/Solution.kt
package study.problem.kt

class Solution {
    fun solution(numbers: IntArray, target: Int): Int = dfs(numbers, target)

    private fun dfs(numbers: IntArray, target: Int): Int =
        numbers.fold(listOf(0)) { list, number ->
            list.run {
                map { it + number } + map { it - number }
            }
        }.count { it == target }
}